### PR TITLE
feat(payments_v2): add finish redirection endpoint

### PR DESCRIPTION
**Pull request by Narayanbhat166**
## Type of Change

- [x] New feature

## Description

This PR adds a new endpoint called `finish_redirection` to payments v2.

## Motivation and Context

Finish redirection endpoint is hit by the processor once the redirection / authentication is performed. 

## How did you test it?

- Create a payment
```bash
curl --location 'http://localhost:8080/v2/payments/create-intent' \
--header 'api-key: dev_cNekK3F8YFiugPANtfmCEe1UpYOP05C71OOMLaCNYB5H1RYtZWuphmKgPZc9aqVi' \
--header 'Content-Type: application/json' \
--header 'x-profile-id: pro_Va3FwgfK5v2F1ZiK3N9r' \
--data '{
    "amount_details": {
        "order_amount": 100,
        "currency": "USD"
    }
}'
```
```json
{
    "id": "12345_pay_019323d64de67fe3b7e62c45751a37cd",
    "status": "requires_payment_method",
    "amount_details": {
        "order_amount": 100,
        "currency": "USD",
        "shipping_cost": null,
        "order_tax_amount": null,
        "skip_external_tax_calculation": "Skip",
        "skip_surcharge_calculation": "Skip",
        "surcharge_amount": null,
        "tax_on_surcharge": null
    },
    "client_secret": "12345_pay_019323d64de67fe3b7e62c45751a37cd_secret_019323d64de67fe3b7e62c53b39411be",
    "merchant_reference_id": null,
    "routing_algorithm_id": null,
    "capture_method": "automatic",
    "authentication_type": "no_three_ds",
    "billing": null,
    "shipping": null,
    "customer_id": null,
    "customer_present": "Present",
    "description": null,
    "return_url": null,
    "setup_future_usage": "on_session",
    "apply_mit_exemption": "Skip",
    "statement_descriptor": null,
    "order_details": null,
    "allowed_payment_method_types": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "payment_link_enabled": "Skip",
    "payment_link_config": null,
    "request_incremental_authorization": "default",
    "expires_on": "2024-11-13T04:59:27.494Z",
    "frm_metadata": null,
    "request_external_three_ds_authentication": "Skip"
}
```

- Confirm the payment with a 3ds card
```bash
curl --location 'http://localhost:8080/v2/payments/12345_pay_019323d64de67fe3b7e62c45751a37cd/confirm-intent' \
--header 'x-profile-id: pro_Va3FwgfK5v2F1ZiK3N9r' \
--header 'x-client-secret: 12345_pay_019323d64de67fe3b7e62c45751a37cd_secret_019323d64de67fe3b7e62c53b39411be' \
--header 'Content-Type: application/json' \
--header 'api-key: pk_dev_72ad688781744773b42030b6b7000e80' \
--data '{
    "payment_method_data": {
        "card": {
            "card_number": "4000008400001280",
            "card_exp_month": "01",
            "card_exp_year": "25",
            "card_holder_name": "John Doe",
            "card_cvc": "100"
        }
    },
    "payment_method_type": "card",
    "payment_method_subtype": "credit",
    "client_secret": "33333_pay_0192570fe84d7c02bf4df06c650e216e_secret_0192570fe85773c1916af7b3288bd8f1"
}'
```

```json
{
    "id": "12345_pay_019323d64de67fe3b7e62c45751a37cd",
    "status": "requires_customer_action",
    "amount": {
        "order_amount": 100,
        "currency": "USD",
        "shipping_cost": null,
        "order_tax_amount": null,
        "skip_external_tax_calculation": "Skip",
        "skip_surcharge_calculation": "Skip",
        "surcharge_amount": null,
        "tax_on_surcharge": null,
        "net_amount": 100,
        "amount_to_capture": null,
        "amount_capturable": 0,
        "amount_captured": null
    },
    "connector": "stripe",
    "client_secret": "12345_pay_019323d64de67fe3b7e62c45751a37cd_secret_019323d64de67fe3b7e62c53b39411be",
    "created": "2024-11-13T04:44:27.494Z",
    "payment_method_data": null,
    "payment_method_type": "card",
    "payment_method_subtype": "credit",
    "next_action": {
        "type": "redirect_to_url",
        "redirect_to_url": "http://localhost:8080/v2/payments/12345_pay_019323d64de67fe3b7e62c45751a37cd/start_redirection?publishable_key=pk_dev_72ad688781744773b42030b6b7000e80&profile_id=pro_Va3FwgfK5v2F1ZiK3N9r"
    },
    "connector_transaction_id": "pi_3QKYROIVaesDjvMP1uVuUQYg",
    "connector_reference_id": null,
    "merchant_connector_id": "mca_mcB1x1Sg0OaPyc0Efywe",
    "browser_info": null,
    "error": null
}
```

- Complete the redirection
image

## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by jarnura — APPROVED** [review]

File: crates/common_utils/src/types.rs [hunk 1]
```diff
@@ -1516,3 +1529,54 @@ pub trait ConnectorTransactionIdTrait {
         self.get_optional_connector_transaction_id()
     }
 }
+
+/// Domain type for Statement Descriptor
```
Comment (sai-harsha-vardhan) [inline crates/common_utils/src/types.rs:1533]:
```suggestion
/// Domain type for PublishableKey
```

File: crates/hyperswitch_domain_models/src/payments.rs [hunk 2]
```diff
@@ -129,6 +130,24 @@ impl PaymentIntent {
...
+    ) -> CustomResult<url::Url, errors::api_error_response::ApiErrorResponse> {
+        let finish_redirection_url = format!(
+            "{base_url}/v2/payments/{}/finish_redirection/{publishable_key}/{}",
+            self.id.get_string_repr(),
+            self.profile_id.get_string_repr()
+        );
+
+        url::Url::parse(&finish_redirection_url)
+            .change_context(errors::api_error_response::ApiErrorResponse::InternalServerError)
+            .attach_printable("Error creating start redirection url")
```
Comment (sai-harsha-vardhan) [inline crates/hyperswitch_domain_models/src/payments.rs:149]:
```suggestion
            .attach_printable("Error creating finish redirection url")
```

File: crates/router/src/core/payments.rs [hunk 3]
```diff
@@ -1893,6 +1951,167 @@ impl PaymentRedirectFlow for PaymentRedirectSync {
...
+            | common_enums::IntentStatus::RequiresMerchantAction
+            | common_enums::IntentStatus::RequiresCapture
+            | common_enums::IntentStatus::PartiallyCaptured
+            | common_enums::IntentStatus::RequiresConfirmation
+            | common_enums::IntentStatus::PartiallyCapturedAndCapturable => {
+                Err(errors::ApiErrorResponse::PaymentUnexpectedState {
+                    current_flow: format!("{self:?}"),
+                    field_name: "status".to_string(),
+                    current_value: intent_status.to_string(),
+                    states: ["requires_payment_method".to_string()].join(", "),
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/payments.rs:1976]:
```suggestion
                    states: ["requires_customer_action".to_string()].join(", "),
```

**Review by sai-harsha-vardhan — DISMISSED** [review]

**Review by ShivanshMathurJuspay — DISMISSED** [review]

File: crates/router/src/core/user.rs [hunk 4]
```diff
@@ -2190,6 +2190,7 @@ pub async fn list_user_authentication_methods(
     ))
 }
 
+#[cfg(feature = "v1")]
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/router/src/core/user.rs:2193]: the jsonRedirectiontype has changed, so had to flag this under v1

**Review by ThisIsMani — DISMISSED** [review]

File: crates/router/src/routes/dummy_connector.rs [hunk 5]
```diff
@@ -13,6 +13,7 @@ mod errors;
 pub mod types;
 mod utils;
 
+#[cfg(all(feature = "dummy_connector", feature = "v1"))]
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/router/src/routes/dummy_connector.rs:16]: If we decide to have a dummy connector for v2, we can remove the feature flags and test, or migrate them later

**Review by sai-harsha-vardhan — DISMISSED** [review]

**Review by jarnura — DISMISSED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Review by ThisIsMani — APPROVED** [review]
LGTM

**Review by ShivanshMathurJuspay — APPROVED** [review]

**Pull request merged**